src/webgpu/shader/types.ts

```diff
@@ -212,33 +261,43 @@ export function* generateTypes({
 
   // Array types
   if (containerType === 'array') {
+    // Buffer affective binding size must be a multiple of 4. Adjust array length as needed.
+    let arrayLength = kDefaultArrayLength;
+    if (
+      addressSpace === 'storage' &&
+      scalarInfo.layout !== undefined &&
+      scalarInfo.layout.alignment % 4 !== 0
+    ) {
+      arrayLength = align(arrayLength, 4);
```

Comment:
I think this can make arrayLength bigger than it should be. For example for f16, scalarInfo.layout.alignment is 2.
If arrayLength is 6, the byte size will be 12. That's already a multiple of 4.
But this increases arrayLength to 8, making the byte size 16.
(I am not sure if this is a big problem, or just a nit.)

```suggestion
      scalarInfo.layout.alignment < 4
    ) {
      arrayLength = align(arrayLength, 4 / scalarInfo.layout.alignment);
```

The need for the `addressSpace === 'storage'` condition is also surprising until you dig into what happens later with arrayStride(). Since this setup has grown more complex I think it would be best to rework the whole block for readability:
```ts
  if (containerType === 'array') {
    let arrayLength = kArrayLength;
    let arrayLayout = undefined;
    if (scalarInfo.layout) {
      // Uniform storage class must have array elements aligned to 16.
      const elementAlignment = addressSpace === 'uniform' ? 16 : scalarInfo.layout.alignment;

      // Buffer's effective binding size must be a multiple of 4.
      if (elementAlignment < 4) {
        arrayLength = align(arrayLength, 4 / elementAlignment);
      }

      const arrayByteSize = kDefaultArrayLength * align(scalarInfo.layout.size, elementAlignment);
      arrayLayout = { alignment: scalarInfo.layout.alignment, size: arrayByteSize };
    }

    const arrayTypeInfo = { elementBaseType: `${baseType}`, arrayLength, layout: arrayLayout };
```

Reply:
I totally reworked this.  Should have much better documentation about what the outputs should be.